doc/dns-resovler.md

```diff
@@ -0,0 +1,58 @@
+
+Also we like to use nginx facilities for load balancing and proxying. That means we have to use `balancer_by_lua` directive which can set the load balancer. But it has to set it to the IP address, not the DNS name. So we have to resolve that DNS name into IP address and there is no API exposed in nginx to do that.
+
+## Requirements
+
+Our DNS resolver should meet following criteria:
+
+* caching for duration of TTL
+* resolve A and CNAME records
+* using stale cache when TTL expires and cache can't be updated
+* many parallel requests to one worker result in one query
+* connect to multiple DNS servers in parallel
+* working search scope
+
+## Scenarios
+
+### Low traffic
+
+A request is made to APIcast and DNS cache is updated. Then there is no request for TTL of that record and another request is made after that record expires. Should APIcast try to use stale cache or hold the request? For how long?
+
+### High traffic (on boot)
+
+1000 requests are made in the same time. First that arrives starts resolving DNS and others should wait until it is completed. After that request is completed, other requests are going to use that result.
+
+If first request timeouts, next one should try to do another DNS query and other requests should wait for it to complete.
+
+### High traffic (with cache)
+
+With constant traffic 1000 requests per second when the DNS cache expires should we use stale cache before it is refreshed? Or should requests wait for it to be updated?
+
+## Design
+
+Using OpenResty's `lua-resty-dns` we can connect to any DNS server and implement our own resolver.
+
+DNS resolver caches each record individually and fetches it from the cache recursively. That allows us to have layers of cache and fill just the missing layers. Example: CNAME (TTL 3600) -> CNAME (TTL 360) -> A (TTL 60) will just query the A record when it expires.
+
+Querying multiple nameservers in parallel is accomplished by using "light thread". The first nameserver that responds is used.
+
+De-duplicating queries is achieved by using ngx.semaphore to wait for already executing queries. The waiting algorithm can be very much improved as right now it just waits for any query being executed, not the specific one. Stale cache is being returned when timeout is reached and query can't be completed on time. Queries are performed in the background timer loop.
+
+Search scope defined in `/etc/resolv.conf` is evaluted one by one until matching record is returned. This could be optimized by executing all queries in parallel.
+
+Following rules apply to handle scenarios described above:
+
+* When cache is missing, all incoming requests de-duplicate queries and wait for the first one to complete. If does not complete on time, the next one tries again, ...
+* When cache is stale, use stale record and trigger de-duplicated query in the background.
+* If record can't be updated, use stale cache for unlimited time, but try to recover.
```

Comment:
Add to this: sound the alarm and transition to red alert, make noise, eventually drop the stale cache and start returning 500s. Admins should know ASAP, and if not taken care of, things should start failing hard (instead of, for example, partially failing because some host not being available is not a fatal error but a subtle one). I guess there are a lot of different possible behaviours, perhaps configuration or overriding of default behaviour is desirable.